builtin/client/chatcommands.lua

```diff
@@ -34,3 +34,55 @@ core.register_on_sending_chat_messages(function(message)
 
 	return true
 end)
+
+core.register_chat_autocompletion(function(message, cursorpos, first_invocation)
+	if message:sub(1, 1) ~= "/" then
+		return
+	end
+	local firstspace = message:find(" ")
+	if firstspace then
+		local cmd = message:sub(2, firstspace-1)
+		local def = core.registered_chatcommands[cmd]
+		if not def then
+			core.display_chat_message("Invalid chatcommand \"" .. cmd .. "\".")
+			return
+		end
+		if not def.autocompletion then
+			return
+		end
+		local params = message:sub(firstspace + 1)
+		local cursorpos_rel = cursorpos - firstspace
+		local newmsg, newcp, abort = def.autocompletion(params,	cursorpos_rel,
+				first_invocation)
+		if type(newmsg) == "string" then
+			newmsg = "/" .. cmd .. " " .. newmsg
+		end
+		if type(newcp) == "number" then
+			newcp = newcp + firstspace
+		end
+		return newmsg, newcp, abort
+	end
+
+	local cmdstart = message:sub(2)
+	local replacement, possible_cmds = core.autocomplete_word(cmdstart,
+			core.registered_chatcommands)
+
+	if not replacement then
+		core.display_chat_message("No chatcommand found beginning with \"" ..
```

Comment:
Maybe it would be a good idea to use `fgettext` for all messages shown to the user?